File: www/flex/features/auth/views/login_view.rs
// ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
// ┃ Copyright: (c) 2023, Mike 'PhiSyX' S. (https://github.com/PhiSyX)         ┃
// ┃ SPDX-License-Identifier: MPL-2.0                                          ┃
// ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
// ┃                                                                           ┃
// ┃  This Source Code Form is subject to the terms of the Mozilla Public      ┃
// ┃  License, v. 2.0. If a copy of the MPL was not distributed with this      ┃
// ┃  file, You can obtain one at https://mozilla.org/MPL/2.0/.                ┃
// ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

use flex_web_framework::{html, Node, SessionFlashExtension, ViewInterface};

use crate::features::auth::errors::LoginError;
use crate::features::auth::forms::LoginFormData;
use crate::features::auth::responses::CreationAccountReply;
use crate::features::auth::routes::web::AuthRouteID;
use crate::templates::layouts::BaseHTMLLayout;

// --------- //
// Structure //
// --------- //

#[derive(Default)]
#[derive(flex_web_framework::View)]
pub struct LoginView
{
	pub identifier: String,
	pub password: String,
	pub remember_me: bool,
	pub success_message: Option<CreationAccountReply>,
	pub error_message: Option<LoginError>,
}

// -------------- //
// Implémentation // -> Interface
// -------------- //

impl ViewInterface for LoginView
{
	type Layout = BaseHTMLLayout;
	type Metadata = Node;
	type Scripts = Node;
	type Styles = Node;
	type View = Node;

	async fn with_session(
		mut self,
		session: &flex_web_framework::sessions::Session,
	) -> Self
	{
		self.success_message = session.take(CreationAccountReply::KEY).await;
		self.error_message = session.take(LoginError::KEY).await;
		self
	}

	fn title(&self) -> impl ToString
	{
		"Auth - Connexion"
	}

	fn styles(&self) -> Self::Styles
	{
		html! {
			<link href="/public/css/auth.css" rel="stylesheet">
		}
	}

	fn view(&self) -> Self::View
	{
		html! {
			<div id="no-auth-intro">
				<div class="auth-container">
					<h1 class="auth-heading">Connexion</h1>

					<div
						let-success:option={&self.success_message}
						class="alert alert/success [ gap=1 ]"
					>
						<div class="[ flex:full ]">{success}</div>
					</div>

					<div
						let-error:option={&self.error_message}
						class="alert alert/error [ gap=1 ]"
					>
						<div class="[ flex:full ]">{error}</div>
					</div>

					<form action={AuthRouteID::Login} method="POST">
						<div class="form-group">
							<label for="identifier">"Identifiant"</label>
							<input
								id="identifier"
								name="identifier"
								value={&self.identifier}
								placeholder="Nom d'utilisateur ou adresse e-mail"
								required
								tabindex="1"
							>
						</div>

						<div class="form-group">
							<span class="flex gap=2 space-between">
								<label for="password">"Mot de passe"</label>
								<a
									href="#"
									//href={AuthRouteID::Forget}
								>
									"Mot de passe oublié ?"
								</a>
							</span>

							<input
								id="password"
								name="password"
								type="password"
								value={&self.password}
								required
								tabindex="2"
							>
						</div>

						<div class="flex gap=2 space-between">
							<label>
								"Connexion automatique lors des prochaines sessions"
							</label>

							<span class="flex gap=1">
								<label for="yes">
									"Oui "
									<input
										id="yes"
										type="radio"
										name="remember_me"
										checked:if=(self.remember_me)
										value="true"
										tabindex="3"
									>
								</label>

								<label for="no">
									<input
										id="no"
										type="radio"
										name="remember_me"
										checked:if=(!self.remember_me)
										value="false"
										tabindex="4"
									>
									" Non"
								</label>
							</span>
						</div>

						<div class="form-group">
							<button class="btn-submit" type="submit" tabindex="5">
								"Se connecter"
							</button>
						</div>
					</form>

					<hr text="OU">

					<a
						href={AuthRouteID::Signup}
						class="d-block btn-submit t-center"
					>
						"Aller à la page d'inscription"
					</a>
				</div>

			</div>
		}
	}
}

impl From<LoginFormData> for LoginView
{
	fn from(form: LoginFormData) -> Self
	{
		Self {
			identifier: form.identifier.to_string(),
			password: form.password,
			remember_me: form.remember_me,
			success_message: Default::default(),
			error_message: Default::default(),
		}
	}
}

File: apps/web/app-scratch/src/views/login/login-view.ts
// ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
// ┃ Copyright: (c) 2024, Mike 'PhiSyX' S. (https://github.com/PhiSyX)         ┃
// ┃ SPDX-License-Identifier: MPL-2.0                                          ┃
// ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
// ┃                                                                           ┃
// ┃  This Source Code Form is subject to the terms of the Mozilla Public      ┃
// ┃  License, v. 2.0. If a copy of the MPL was not distributed with this      ┃
// ┃  file, You can obtain one at https://mozilla.org/MPL/2.0/.                ┃
// ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

import { customElement } from "@phisyx/flex-custom-element";
import {
	div,
	form,
	h1,
	label,
	main,
	section,
	span,
} from "@phisyx/flex-html-element-extension";

import { signal } from "@phisyx/flex-signal";

import {
	MAXLENGTH_NICKNAME,
	PLACEHOLDER_NICKNAME,
	RememberMeStorage,
	VALIDATION_NICKNAME_INFO,
} from "@phisyx/flex-chat";
import {
	buttonIcon,
	inputSwitch,
	textInput,
	uiButton,
} from "@phisyx/flex-uikit";
import scss from "./login-view.scss?url";

@customElement({ mode: "open", styles: [scss] })
export default class LoginView {
	declare static TAG_NAME: string;

	advancedInfo = signal(false);
	websocketServerURL = signal("");
	passwordServer = signal("");
	nickname = signal("");
	alternativeNickname = signal("");
	realname = signal("");
	channels = signal("");
	rememberMe = new RememberMeStorage();
	loader = signal(false);

	render() {
		return main(
			section(
				h1("Accès direct au Chat"),

				form(
					textInput(
						this.websocketServerURL,
						{
							label: "url",
							name: "server",
						},
						{
							placeholder: "URL WebSocket du serveur de Chat",
							type: "url",
						},
					)
						.displayWhen(this.advancedInfo)
						// FIXME: trouver un moyen d'automatiser cette partie.
						.on("sync:model", (evt: CustomEvent<string>) => {
							this.websocketServerURL.set(evt.detail);
						}),

					textInput(
						this.passwordServer,
						{
							label: "password",
							name: "password_server",
						},
						{
							placeholder: "Mot de passe du serveur de Chat",
							type: "password",
						},
					)
						.displayWhen(this.advancedInfo)
						// FIXME: trouver un moyen d'automatiser cette partie.
						.on("sync:model", (evt: CustomEvent<string>) => {
							this.passwordServer.set(evt.detail);
						}),

					textInput(
						this.nickname,
						{
							label: "user",
							name: "nickname",
							// error: errors.nickname
						},
						{
							maxLength: MAXLENGTH_NICKNAME,
							placeholder: PLACEHOLDER_NICKNAME,
						},
					)
						.title(VALIDATION_NICKNAME_INFO)
						// FIXME: trouver un moyen d'automatiser cette partie.
						.on("sync:model", (evt: CustomEvent<string>) => {
							this.nickname.set(evt.detail);
						}),

					textInput(
						this.alternativeNickname,
						{
							label: "user",
							name: "alternative_nickname",
						},
						{
							maxLength: MAXLENGTH_NICKNAME,
							placeholder: "Pseudonyme alternatif",
						},
					)
						.displayWhen(this.advancedInfo)
						.title(VALIDATION_NICKNAME_INFO)
						// FIXME: trouver un moyen d'automatiser cette partie.
						.on("sync:model", (evt: CustomEvent<string>) => {
							this.alternativeNickname.set(evt.detail);
						}),

					textInput(
						this.realname,
						{
							label: "user",
							name: "realname",
						},
						{
							placeholder: "Nom réel",
						},
					)
						.displayWhen(this.advancedInfo)
						// FIXME: trouver un moyen d'automatiser cette partie.
						.on("sync:model", (evt: CustomEvent<string>) => {
							this.alternativeNickname.set(evt.detail);
						}),

					textInput(
						this.channels,
						{
							label: "channel",
							name: "channels",
						},
						{
							placeholder: "Salons à rejoindre",
						},
					)
						// FIXME: trouver un moyen d'automatiser cette partie.
						.on("sync:model", (evt: CustomEvent<string>) => {
							this.channels.set(evt.detail);
						}),
				)
					.id("#chat-login-form")
					.class("ov:h flex! border/radius=1")
					.action("/chat/login")
					.method("POST")
					.submitWith({
						success: this.submitHandler,
					}),
				div(
					buttonIcon("plus")
						.title("Afficher les champs avancés")
						.onClick(this.displayAdvancedInfoHandler),
				)
					.class("align-t:center")
					.displayWhen(this.advancedInfo.computed((b) => !b)),
				div(
					label(
						"Connexion automatique lors de vos prochaines sessions :",
					),
					inputSwitch(this.rememberMe.value, {
						labelN: "Non",
						labelY: "Oui",
						name: "remember_me",
					}).on("sync:model", (evt: CustomEvent<boolean>) => {
						this.rememberMe.set(evt.detail);
					}),
				).class("remember-me [ m:a align-t:center w=35 ]"),

				this.loader.computed((loader) =>
					uiButton(
						{
							icon: loader ? "loader" : undefined,
							position: "right",
							type: "submit",
						},
						{
							// @ts-expect-error à corriger
							form: "chat-login-form",
						},
						span("Accéder au Chat").class("flex:full"),
					).class("flex align-jc:se p=2 b:none cursor:pointer"),
				),
			).class("flex! gap=3 min-w=43"),
			// 	<UiButton icon="settings" class="color-scheme" @click="toSettingsView" />
			//
			// <ModulesProgress />
		)
			.id("#chat-login-view")
			.class("scroll:y flex! flex/center:full m:a pos-r");
	}

	/**
	 * Affiche les informations de connexion avancées.
	 */
	displayAdvancedInfoHandler = () => {
		this.advancedInfo.set(true);
	};

	submitHandler = (evt: SubmitEvent) => {
		console.log(this, evt);
	};
}
